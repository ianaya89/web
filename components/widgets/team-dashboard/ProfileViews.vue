<template>
  <div class="w-full bg-white shadow-xl p-4">
    <div class="w-full flex">
      <div class="mr-auto">
        <label-chart :label="selectedViews" :info="$t('app.team.dashboard.profile_views')" />
      </div>

      <div class="ml-auto my-auto">
        <growth class="text-h3 font-medium" :icon-class="'text-h4'" :value="selectedGrowth" />
      </div>
    </div>

    <div class="w-full">
      <timeseries-chart
        :series="viewsSeries"
        :height="200"
        :series-type="'areaspline'"
        :mouse-over-handler="growthMouseOver"
        :plot-lines="false"
        :series-color="'#21CE99'"
        :tooltip="false"
        :green-marker="true"
        :fill-color="[[0, 'rgba(33, 206, 153,0.3)'], [1, 'rgba(33, 206, 153,0)']]"
      />
    </div>
  </div>
</template>

<script>
import LabelChart from '@/components/LabelChart'
import Growth from '@/components/Growth'
import TimeseriesChart from '@/components/charts/TimeseriesChart'
import { UnitProfileViewsQuery } from '@/services/graphql'
import { calculateGrowth } from '@/services/helpers'

export default {
  name: 'ProfileViews',

  components: {
    Growth,
    LabelChart,
    TimeseriesChart
  },

  props: {
    unitId: {
      type: String,
      default: null
    }
  },

  apollo: {
    unit: {
      ...UnitProfileViewsQuery,
      variables() {
        return {
          id: this.unitId,
          page: 'PROFILE'
        }
      }
    }
  },

  data: () => ({
    selectedGrowthIndex: null,
    unit: UnitProfileViewsQuery.mock
  }),

  computed: {
    viewsSeries() {
      return {
        name: '',
        data: this.$get(this.unit, 'entity.pageVisits', []).map(item => ({
          x: new Date(item.date),
          y: item.count
        }))
      }
    },

    selectedViews() {
      if (this.viewsSeries.data.length > 0) {
        if (this.selectedGrowthIndex === null) {
          return this.viewsSeries.data[this.viewsSeries.data.length - 1].y
        } else {
          return this.viewsSeries.data[this.selectedGrowthIndex].y
        }
      } else {
        return 0
      }
    },

    selectedGrowth() {
      if (this.viewsSeries.data.length > 0) {
        if (this.selectedGrowthIndex === null) {
          const lastIndex = this.viewsSeries.data.length - 1
          if (lastIndex === 0) {
            return 0
          }
          return calculateGrowth(
            this.viewsSeries.data[lastIndex - 1].y,
            this.viewsSeries.data[lastIndex].y
          )
        } else {
          if (this.selectedGrowthIndex === 0) {
            return 0
          }
          return calculateGrowth(
            this.viewsSeries.data[this.selectedGrowthIndex - 1].y,
            this.viewsSeries.data[this.selectedGrowthIndex].y
          )
        }
      } else {
        return 0
      }
    }
  },

  methods: {
    growthMouseOver(data) {
      this.selectedGrowthIndex = data.target.index
    }
  }
}
</script>

<style scoped>
</style>
