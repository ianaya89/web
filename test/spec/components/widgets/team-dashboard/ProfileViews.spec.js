import { shallowMount } from '@vue/test-utils'
import ProfileViews from '@/components/widgets/team-dashboard/ProfileViews'

describe('ProfileViews', () => {
  it('renders ProfileViews', () => {
    const wrapper = shallowMount(ProfileViews, {
      propsData: {
        unitId: 1
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('returns view series', () => {
    const wrapper = shallowMount(ProfileViews, {
      propsData: {
        unitId: 1
      }
    })

    wrapper.vm.unit = {
      entity: {
        pageVisits: [
          {
            date: '2019-08-05',
            count: 3
          }
        ]
      }
    }

    expect(wrapper.vm.viewsSeries).toEqual({
      name: '',
      data: [
        {
          x: new Date('2019-08-05'),
          y: 3
        }
      ]
    })
  })

  it('returns selected views', () => {
    const wrapper = shallowMount(ProfileViews, {
      propsData: {
        unitId: 1
      }
    })

    wrapper.vm.unit = {
      entity: {
        pageVisits: []
      }
    }

    expect(wrapper.vm.selectedViews).toEqual(0)

    wrapper.vm.unit = {
      entity: {
        pageVisits: [
          {
            date: '2019-08-04',
            count: 8
          },
          {
            date: '2019-08-05',
            count: 3
          }
        ]
      }
    }

    expect(wrapper.vm.selectedViews).toEqual(3)

    wrapper.vm.selectedGrowthIndex = 0

    expect(wrapper.vm.selectedViews).toEqual(8)
  })

  it('it select growth index from mouseover data object', () => {
    const wrapper = shallowMount(ProfileViews, {
      propsData: {
        unitId: 1
      }
    })

    expect(wrapper.vm.selectedGrowthIndex).toEqual(null)

    wrapper.vm.growthMouseOver({
      target: {
        index: 3
      }
    })

    expect(wrapper.vm.selectedGrowthIndex).toEqual(3)
  })
})
