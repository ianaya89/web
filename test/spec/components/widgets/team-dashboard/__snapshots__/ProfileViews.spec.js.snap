// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileViews renders ProfileViews 1`] = `
<div
  class="w-full bg-white shadow-xl p-4"
>
  <div
    class="w-full flex"
  >
    <div
      class="mr-auto"
    >
      <label-chart-stub
        animation="true"
        info="TRANSLATE:app.team.dashboard.profile_views"
        label="0"
        labelclass="text-h3"
        piecolors="#21CE99,#D4D4D4"
      />
    </div>
     
    <div
      class="ml-auto my-auto"
    >
      <growth-stub
        applycolorontext="true"
        class="text-h3 font-medium"
        downcolor="text-red"
        downicon="fas fa-arrow-alt-circle-down"
        duration="2000"
        iconclass="text-h4"
        mainclasses="flex items-center"
        showvalue="true"
        upcolor="text-green"
        upicon="fas fa-arrow-alt-circle-up"
        value="0"
      />
    </div>
  </div>
   
  <div
    class="w-full"
  >
    <timeseries-chart-stub
      enablemousetracking="true"
      fillcolor="0,rgba(33, 206, 153,0.3),1,rgba(33, 206, 153,0)"
      greenmarker="true"
      height="200"
      labelformatter="vm => function () {
        return \`
            <div class=\\"chart__title\\">\${vm.$filters.dateTime(this.point.x)}</div>
            <div class=\\"chart__val\\">\${this.point.filterFnc ? this.point.filterFnc(this.y) : vm.$filters.number(this.y)}</div>
            \`;
      }"
      mouseoverhandler="function () { [native code] }"
      series="[object Object]"
      seriescolor="#21CE99"
      seriestype="areaspline"
    />
  </div>
</div>
`;
